File: edu-ios/Sources/Views/BadgeMedallion.swift
import SwiftUI

// Category/tier metadata for badges (mirrors the web gamification catalog). The
// Badge model only carries `category`/`tier` as strings; this maps them to
// labels, glyphs, and the metal palette used to mint each medallion.
enum BadgeMeta {
    static let categoryOrder = ["milestones", "mastery", "completion", "consistency", "exploration", "special"]

    static func categoryLabel(_ key: String) -> String {
        switch key {
        case "milestones": return "Milestones"
        case "mastery": return "Mastery"
        case "completion": return "Completion"
        case "consistency": return "Consistency"
        case "exploration": return "Exploration"
        case "special": return "Special"
        default: return key.prefix(1).uppercased() + key.dropFirst()
        }
    }

    // SF Symbol engraved on the medallion for each category.
    static func categorySymbol(_ key: String) -> String {
        switch key {
        case "milestones": return "figure.run"
        case "mastery": return "target"
        case "completion": return "building.columns.fill"
        case "consistency": return "flame.fill"
        case "exploration": return "safari.fill"
        case "special": return "sparkles"
        default: return "star.fill"
        }
    }

    static func categoryBlurb(_ key: String) -> String {
        switch key {
        case "milestones": return "Volume — the steady climb."
        case "mastery": return "Quality — proof you learned it."
        case "completion": return "Finishing what you start."
        case "consistency": return "Showing up, day after day."
        case "exploration": return "Breadth across subjects."
        case "special": return "Seasonal, rare, and just for fun."
        default: return ""
        }
    }

    static let tierRank: [String: Int] = ["bronze": 0, "silver": 1, "gold": 2, "platinum": 3, "omniscient": 4]

    static func tierLabel(_ t: String) -> String {
        t == "omniscient" ? "Omniscient" : t.prefix(1).uppercased() + t.dropFirst()
    }

    // Metal face gradient (highlight → shadow) per tier.
    static func tierMetal(_ t: String) -> [Color] {
        switch t {
        case "bronze":     return [Color(hex: 0xe6b877), Color(hex: 0x8f5f2c)]
        case "silver":     return [Color(hex: 0xf2f2f6), Color(hex: 0x96969f)]
        case "gold":       return [Color(hex: 0xf4da8a), Color(hex: 0xb8860b)]
        case "platinum":   return [Color(hex: 0xeaf5ff), Color(hex: 0x9dc6e6)]
        case "omniscient": return [Color(hex: 0xf8ecb8), Color(hex: 0xddb954)]
        default:           return [Color(hex: 0xf4da8a), Color(hex: 0xb8860b)]
        }
    }

    static func tierGlow(_ t: String) -> Color {
        switch t {
        case "gold", "omniscient": return Color(hex: 0xddb954)
        case "platinum":           return Color(hex: 0x9dc6e6)
        default:                   return Color(hex: 0xb8860b)
        }
    }
}

// A procedurally-minted medallion: a metal face (tier gradient) inside a shimmering
// gold rim, with the category symbol engraved in the center. Locked badges become a
// dark disc with a padlock. No image assets.
struct BadgeMedallion: View {
    let badge: Badge
    var size: CGFloat = 68

    var body: some View {
        let unlocked = badge.unlocked
        let face = unlocked
            ? BadgeMeta.tierMetal(badge.tier)
            : [Color(hex: 0x3a2c2c), Color(hex: 0x201414)]
        let rim = unlocked
            ? [Color(hex: 0xf8ecb8), Color(hex: 0xb8860b), Color(hex: 0xf8ecb8), Color(hex: 0x9a7209), Color(hex: 0xf8ecb8)]
            : [Theme.line, Theme.crimson800, Theme.line]

        ZStack {
            Circle()
                .fill(RadialGradient(colors: face, center: UnitPoint(x: 0.35, y: 0.3),
                                     startRadius: 1, endRadius: size * 0.78))
            Circle()
                .strokeBorder(AngularGradient(colors: rim, center: .center),
                              lineWidth: size * 0.075)
            Image(systemName: unlocked ? BadgeMeta.categorySymbol(badge.category) : "lock.fill")
                .font(.system(size: size * 0.38, weight: .bold))
                .foregroundStyle(unlocked ? Color(hex: 0x2a1a0e).opacity(0.82) : Theme.inkSoft)
        }
        .frame(width: size, height: size)
        .shadow(color: BadgeMeta.tierGlow(badge.tier).opacity(unlocked ? 0.5 : 0),
                radius: unlocked ? size * 0.13 : 0)
        .opacity(unlocked ? 1 : 0.5)
    }
}

File: edu-android/app/src/main/java/com/timpsonlyceum/lyceum/ui/components/BadgeMedallion.kt
package com.timpsonlyceum.lyceum.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.AccountBalance
import androidx.compose.material.icons.filled.DirectionsRun
import androidx.compose.material.icons.filled.Explore
import androidx.compose.material.icons.filled.GpsFixed
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.timpsonlyceum.lyceum.model.Badge
import com.timpsonlyceum.lyceum.ui.theme.Theme

/**
 * Category and tier metadata for badges, mirroring the web gamification catalogue.
 *
 * The [Badge] model only carries `category` and `tier` as strings; this maps them
 * to labels, glyphs, and the metal palette each medallion is minted from.
 */
object BadgeMeta {
    val categoryOrder = listOf("milestones", "mastery", "completion", "consistency", "exploration", "special")

    fun categoryLabel(key: String): String = when (key) {
        "milestones" -> "Milestones"
        "mastery" -> "Mastery"
        "completion" -> "Completion"
        "consistency" -> "Consistency"
        "exploration" -> "Exploration"
        "special" -> "Special"
        else -> key.replaceFirstChar { it.uppercase() }
    }

    /** The glyph struck into the face of the medallion, one per category. */
    fun categoryIcon(key: String): ImageVector = when (key) {
        "milestones" -> Icons.Filled.DirectionsRun
        "mastery" -> Icons.Filled.GpsFixed
        "completion" -> Icons.Filled.AccountBalance
        "consistency" -> Icons.Filled.LocalFireDepartment
        "exploration" -> Icons.Filled.Explore
        "special" -> Icons.Filled.AutoAwesome
        else -> Icons.Filled.Star
    }

    fun categoryBlurb(key: String): String = when (key) {
        "milestones" -> "Volume — the steady climb."
        "mastery" -> "Quality — proof you learned it."
        "completion" -> "Finishing what you start."
        "consistency" -> "Showing up, day after day."
        "exploration" -> "Breadth across subjects."
        "special" -> "Seasonal, rare, and just for fun."
        else -> ""
    }

    val tierRank = mapOf("bronze" to 0, "silver" to 1, "gold" to 2, "platinum" to 3, "omniscient" to 4)

    fun tierLabel(t: String): String =
        if (t == "omniscient") "Omniscient" else t.replaceFirstChar { it.uppercase() }

    /** Metal face gradient, highlight to shadow, per tier. */
    fun tierMetal(t: String): List<Color> = when (t) {
        "bronze" -> listOf(Color(0xFFE6B877), Color(0xFF8F5F2C))
        "silver" -> listOf(Color(0xFFF2F2F6), Color(0xFF96969F))
        "gold" -> listOf(Color(0xFFF4DA8A), Color(0xFFB8860B))
        "platinum" -> listOf(Color(0xFFEAF5FF), Color(0xFF9DC6E6))
        "omniscient" -> listOf(Color(0xFFF8ECB8), Color(0xFFDDB954))
        else -> listOf(Color(0xFFF4DA8A), Color(0xFFB8860B))
    }

    fun tierGlow(t: String): Color = when (t) {
        "gold", "omniscient" -> Color(0xFFDDB954)
        "platinum" -> Color(0xFF9DC6E6)
        else -> Color(0xFFB8860B)
    }
}

/**
 * A procedurally-minted medallion: a metal face in the tier's gradient inside a
 * shimmering gold rim, with the category glyph struck in the centre. Locked
 * badges become a dark disc with a padlock. No image assets — the same approach
 * the iOS app takes, so a new badge needs no artwork.
 */
@Composable
fun BadgeMedallion(badge: Badge, size: Dp = 68.dp) {
    val unlocked = badge.unlocked
    val face = if (unlocked) BadgeMeta.tierMetal(badge.tier)
    else listOf(Color(0xFF3A2C2C), Color(0xFF201414))
    val rim = if (unlocked) listOf(
        Color(0xFFF8ECB8), Color(0xFFB8860B), Color(0xFFF8ECB8),
        Color(0xFF9A7209), Color(0xFFF8ECB8),
    ) else listOf(Theme.line, Theme.crimson800, Theme.line)

    Box(
        Modifier.size(size).alpha(if (unlocked) 1f else 0.5f),
        contentAlignment = Alignment.Center,
    ) {
        Canvas(Modifier.size(size)) {
            val r = this.size.minDimension / 2f
            // The highlight sits up and to the left, so the disc reads as struck
            // metal catching a light rather than a flat filled circle.
            drawCircle(
                brush = Brush.radialGradient(
                    colors = face,
                    center = Offset(this.size.width * 0.35f, this.size.height * 0.3f),
                    radius = r * 1.56f,
                ),
                radius = r,
            )
            val rimWidth = this.size.minDimension * 0.075f
            drawCircle(
                brush = Brush.sweepGradient(rim),
                radius = r - rimWidth / 2f,
                style = Stroke(width = rimWidth),
            )
        }
        Icon(
            imageVector = if (unlocked) BadgeMeta.categoryIcon(badge.category) else Icons.Filled.Lock,
            contentDescription = null,
            tint = if (unlocked) Color(0xFF2A1A0E).copy(alpha = 0.82f) else Theme.inkSoft,
            modifier = Modifier.size(size * 0.38f),
        )
    }
}
